reactive/docker_registry.py

```diff
@@ -0,0 +1,221 @@
+
+
+def report_active_status():
+    '''Update status based on related charms/config.'''
+    app_suffix = []
+    if is_flag_set('charm.docker-registry.tls-enabled'):
+        app_suffix.append('tls')
+    else:
+        app_suffix.append('insecure')
+    if is_flag_set('website.available'):
+        app_suffix.append('proxied')
+
+    if app_suffix:
+        status_suffix = ' ({})'.format(', '.join(app_suffix))
+    else:
+        status_suffix = ''
+    layer.status.active('Ready{}.'.format(status_suffix))
+
+
+@when('apt.installed.docker.io')
+@when_not('charm.docker-registry.configured')
+def start():
+    layer.status.maint('Configuring the registry.')
+
+    layer.docker_registry.configure_registry()
+    layer.docker_registry.start_registry()
+
+    set_flag('charm.docker-registry.configured')
+    report_active_status()
+
+
+@when('charm.docker-registry.configured')
+@when_any('config.changed',
+          'leadership.changed.http-secret')
+def config_changed():
+    layer.status.maint('Reconfiguring the registry.')
+    charm_config = hookenv.config()
+    name = charm_config.get('registry-name')
+
+    # If our name changed, make sure we stop the old one
+    if charm_config.changed('registry-name') and charm_config.previous('registry-name'):
+        name = charm_config.previous('registry-name')
+
+    layer.docker_registry.stop_registry(name=name)
+    layer.docker_registry.configure_registry()
+    layer.docker_registry.start_registry()
+
+    report_active_status()
+
+
+@when('charm.docker-registry.configured')
+@when('endpoint.docker-registry.requests-pending')
+def handle_requests():
+    '''Set all the registry config that clients may care about.'''
+    registry = endpoint_from_flag('endpoint.docker-registry.requests-pending')
+    charm_config = hookenv.config()
+    port = charm_config.get('registry-port')
+    http_config = charm_config.get('http_config', '')
+
+    # tls config
+    url_prefix = 'http'
+    tls_enabled = False
+    tls_ca = None
+    if is_flag_set('charm.docker-registry.tls-enabled'):
+        url_prefix = 'https'
+        tls_enabled = True
+        cert_provider = endpoint_from_flag('cert-provider.ca.available')
+        if cert_provider:
+            tls_ca = cert_provider.root_ca_cert
+
+    # http config
+    if http_config and not http_config == "":
+        # When set, this config is our source of truth for the registry url
+        registry_url = http_config
+    else:
+        registry_url = '{}://{}:{}'.format(url_prefix, hookenv.unit_public_ip(), port)
+
+    for request in registry.requests:
+        hookenv.log('Sending {} and tls config to registry client.'.format(registry_url))
+        request.set_registry_config(registry_url=registry_url,
+                                    tls_enabled=tls_enabled,
+                                    tls_ca=tls_ca)
+    registry.mark_completed()
+
+
+@when('cert-provider.available')
+def request_certificates():
```

Comment:
This looks as if we "Request a new cert/key initally, and every 5 minutes check for a sans update".

Reply:
This is true.  We allow people to configure http-host, which could affect the SANs list.  However, it's probably better to have a separate config handler for that and avoid running this every 5 minutes.  I'll investigate that approach.